GVFS/FastFetch/Index.cs

```diff
@@ -213,7 +213,7 @@ private bool UpdateFileInformationFromDiskForFiles(MemoryMappedViewAccessor inde
                     addedOrEditedLocalFiles,
                     (localPath) =>
                     {
-                        string gitPath = FromDotnetFullPathToGitRelativePath(localPath, this.repoRoot);
+                        string gitPath = localPath.Replace(Path.DirectorySeparatorChar, GVFSConstants.GitPathSeparator);

```

Comment:
This seems wrong for two, conflicting reasons:

1. If this was here, then `localPath` needs removing `this.repoRoot`. Is `localPath` already local and `this.repoRoot` empty?
2. There is another instance of `FromDotnetFullPathToGeitRelativePath()` in this file. Why only one changed?

I'm having trouble resolving the dissonance here, as the method context looks like `localPath` shouldn't contain `this.repoRoot`, in which case this would be a bug!

@kivikakk: you were in here, so how did you find that this needed changing? Can you resolve my confusion here?

Reply:
1. The purpose of this change is to exactly make `this.repoRoot` no longer part of these paths. (`this.repoRoot` is never empty.) The problem is that there are multiple repository roots depending on what you want -- the working tree, or the backing store. These are the same on Windows and Mac and therefore it was fine to use absolute paths in many places in the VFS4G codebase, because that absolute path was fit for use in whatever place you might want to, whether that's triggering hydration (working tree) or writing the hydrated file contents to disk (backing store).

   On Linux, these paths are different -- `src/` for the working tree, `.vfsforgit/lower/` for the backing store. Accordingly, we can't meaningfully pass around absolute paths because any given absolute path is only fit for one of those two purposes. I briefly considered continuing to use the absolute working tree path (`src/`) everywhere and replacing the prefix with the backing store path (`.vfsforgit/linux/`) where needed, but while a smaller change, is a huge hack that requires a lot of explanation, and is essentially technical debt that would come to bite us later, as it doesn't require a Windows or Mac programmer to think about which of the two paths are actually desired here.

   Accordingly, this change changes our internal diff objects to use git-relative paths everywhere, and construct appropriate absolute paths at time of use. `addedOrEditedLocalFiles` comes from `DiffHelper` and so has relative paths, which is why we no longer call `FromDotnetFullPathToGitRelativePath`.

2. I originally blindly changed both without looking at the contexts of each, and hit a new error which vexed me for a while. The remaining use is on `file.FullName` where `file` is a `FileInfo` from `FastFetch.WorkingTree.ForAllFiles`. These are absolute paths that come directly from the filesystem and are guaranteed to have a `this.repoRoot` prefix, so we must remove that prefix to correctly index into `this.indexEntryOffsets`.

Reply:
Thanks for the context!